### PR TITLE
Type annotations for LruCache

### PR DESCRIPTION
**Pull request by richvdh**
~~based on #8561.~~

### CODE REVIEW

File: synapse/push/push_rule_evaluator.py [hunk 1]
```diff
@@ -173,19 +173,21 @@ def _contains_display_name(self, display_name: str) -> bool:
...
 
     def _get_value(self, dotted_key: str) -> Optional[str]:
         return self._value_cache.get(dotted_key, None)
 
 
 # Caches (string, is_glob, word_boundary) -> regex for push. See _glob_matches
-regex_cache = LruCache(50000, "regex_push_cache")
+regex_cache = LruCache(
+    50000, "regex_push_cache"
+)  # type: LruCache[Tuple[str, bool, bool],Pattern]
```
Comment (clokep) [inline synapse/push/push_rule_evaluator.py:190]:
```suggestion
)  # type: LruCache[Tuple[str, bool, bool], Pattern]
```

File: synapse/util/caches/lrucache.py [hunk 2]
```diff
@@ -188,8 +206,31 @@ def delete_node(node):
...
+            callbacks: Iterable[Callable[[], None]] = ...,
+            update_metrics: bool = ...,
+        ) -> Union[T, VT]:
+            ...
+
         @synchronized
-        def cache_get(key, default=None, callbacks=[], update_metrics=True):
+        def cache_get(
+            key: KT,
+            default=None,
```
Comment (clokep) [inline synapse/util/caches/lrucache.py:230]: Does making this `Optional[T]` cause mypy to be unhappy?

File: synapse/util/caches/lrucache.py [hunk 3]
```diff
@@ -188,8 +206,31 @@ def delete_node(node):
             node.callbacks.clear()
             return deleted_len
 
+        @overload
+        def cache_get(
+            key: KT,
+            default: Literal[None] = None,
+            callbacks: Iterable[Callable[[], None]] = ...,
+            update_metrics: bool = ...,
```
Comment (clokep) [inline synapse/util/caches/lrucache.py:214]: I did not realize you could use `...` in this situation, that's helpful. 👍

File: synapse/util/caches/lrucache.py [hunk 4]
```diff
@@ -241,8 +282,16 @@ def cache_set_default(key, value):
                 evict()
                 return value
 
+        @overload
+        def cache_pop(key: KT, default: Literal[None] = None) -> Union[None, VT]:
```
Comment (clokep) [inline synapse/util/caches/lrucache.py:286]: I think this is equivalent?

```suggestion
        def cache_pop(key: KT, default: Literal[None] = None) -> Optional[VT]:
```

File: synapse/util/caches/lrucache.py [hunk 5]
```diff
@@ -241,8 +282,16 @@ def cache_set_default(key, value):
...
+            ...
+
+        @overload
+        def cache_pop(key: KT, default: T) -> Union[T, VT]:
+            ...
+
         @synchronized
-        def cache_pop(key, default=None):
+        def cache_pop(key: KT, default=None):
```
Comment (clokep) [inline synapse/util/caches/lrucache.py:294]: Same question as with `cache_get`.

File: synapse/util/caches/lrucache.py [hunk 6]
```diff
@@ -15,12 +15,30 @@
...
+from typing_extensions import Literal
 
 from synapse.config import cache as cache_config
 from synapse.util.caches import CacheMetric, register_cache
 from synapse.util.caches.treecache import TreeCache
 
+T = TypeVar("T")
+FT = TypeVar("FT", bound=Callable[..., Any])
+KT = TypeVar("KT")
+VT = TypeVar("VT")
```
Comment (clokep) [inline synapse/util/caches/lrucache.py:40]: There's enough different types here that these could probably use some comments.

File: synapse/util/caches/lrucache.py (hunk 2 above)
Comment (richvdh) [reply synapse/util/caches/lrucache.py:230]: no, though I'm not sure it does anything.

File: synapse/util/caches/lrucache.py (hunk 5 above)
Comment (richvdh) [reply synapse/util/caches/lrucache.py:294]: same answer :)

**Review by clokep — APPROVED** [review]

**Pull request merged**